README.md
# This repository is deprecated

Each codec is now separate and decoupled from others. The new repositories are:

https://github.com/softwaregroup-bg/ut-codec-ndc

https://github.com/softwaregroup-bg/ut-codec-payshield

https://github.com/softwaregroup-bg/ut-codec-iso8583

https://github.com/softwaregroup-bg/ut-codec-smpp

https://github.com/softwaregroup-bg/ut-codec-jsonrpc

https://github.com/softwaregroup-bg/ut-codec-string


CHANGELOG.md
<a name="5.16.1"></a>
## [5.16.1](https://github.com/softwaregroup-bg/ut-codec/compare/v5.16.0...v5.16.1) (2017-10-23)



<a name="5.16.0"></a>
# [5.16.0](https://github.com/softwaregroup-bg/ut-codec/compare/v5.15.0...v5.16.0) (2017-10-23)


### Bug Fixes

* add dependency ([e162961](https://github.com/softwaregroup-bg/ut-codec/commit/e162961))
* differentiate tags ([#24](https://github.com/softwaregroup-bg/ut-codec/issues/24)) ([d8558b3](https://github.com/softwaregroup-bg/ut-codec/commit/d8558b3))
* improve error message ([d5c5d48](https://github.com/softwaregroup-bg/ut-codec/commit/d5c5d48))
* index out of bounds ([dbd3eb6](https://github.com/softwaregroup-bg/ut-codec/commit/dbd3eb6))
* KW/KQ parsing ([c05d3e1](https://github.com/softwaregroup-bg/ut-codec/commit/c05d3e1))
* missing break ([#23](https://github.com/softwaregroup-bg/ut-codec/issues/23)) ([adedd10](https://github.com/softwaregroup-bg/ut-codec/commit/adedd10))
* separate emv logic ([e0319ef](https://github.com/softwaregroup-bg/ut-codec/commit/e0319ef))
* throw decode error to force disconnect ([44369da](https://github.com/softwaregroup-bg/ut-codec/commit/44369da))
* track transactionRequestId in faults ([b956a6d](https://github.com/softwaregroup-bg/ut-codec/commit/b956a6d))
* use matcher instead of match, which is not recommended ([0ca2595](https://github.com/softwaregroup-bg/ut-codec/commit/0ca2595))


### Features

* fallback to $meta.method ([69d0e62](https://github.com/softwaregroup-bg/ut-codec/commit/69d0e62))
* full header and footer support ([c14f47c](https://github.com/softwaregroup-bg/ut-codec/commit/c14f47c))
* map 0800 network codes ([#27](https://github.com/softwaregroup-bg/ut-codec/issues/27)) ([a96b3da](https://github.com/softwaregroup-bg/ut-codec/commit/a96b3da))
* track transaction  request timeouts ([f6af754](https://github.com/softwaregroup-bg/ut-codec/commit/f6af754))



<a name="5.15.0"></a>
# [5.15.0](https://github.com/softwaregroup-bg/ut-codec/compare/v5.14.0...v5.15.0) (2017-09-13)


### Features

* zmk arqc arpc ([#19](https://github.com/softwaregroup-bg/ut-codec/issues/19)) ([b931d47](https://github.com/softwaregroup-bg/ut-codec/commit/b931d47))



<a name="5.14.0"></a>
# [5.14.0](https://github.com/softwaregroup-bg/ut-codec/compare/v5.13.0...v5.14.0) (2017-09-07)


### Bug Fixes

* separate emv logic ([807522d](https://github.com/softwaregroup-bg/ut-codec/commit/807522d))



<a name="5.13.0"></a>
# [5.13.0](https://github.com/softwaregroup-bg/ut-codec/compare/v5.12.2...v5.13.0) (2017-09-03)


### Features

* emv tags list ([#16](https://github.com/softwaregroup-bg/ut-codec/issues/16)) ([3d33170](https://github.com/softwaregroup-bg/ut-codec/commit/3d33170))



<a name="5.12.2"></a>
## [5.12.2](https://github.com/softwaregroup-bg/ut-codec/compare/v5.12.1...v5.12.2) (2017-08-25)


### Bug Fixes

* separate emv logic ([010869f](https://github.com/softwaregroup-bg/ut-codec/commit/010869f))



<a name="5.12.1"></a>
## [5.12.1](https://github.com/softwaregroup-bg/ut-codec/compare/v5.12.0...v5.12.1) (2017-08-14)


### Bug Fixes

* arqc format ([77ffba3](https://github.com/softwaregroup-bg/ut-codec/commit/77ffba3))



<a name="5.12.0"></a>
# [5.12.0](https://github.com/softwaregroup-bg/ut-codec/compare/v5.11.1...v5.12.0) (2017-08-01)


### Features

* emv parsing ([d85f336](https://github.com/softwaregroup-bg/ut-codec/commit/d85f336))



<a name="5.11.1"></a>
## [5.11.1](https://github.com/softwaregroup-bg/ut-codec/compare/v5.11.0...v5.11.1) (2017-07-28)


### Bug Fixes

* translade -> translate ([8624c30](https://github.com/softwaregroup-bg/ut-codec/commit/8624c30))



<a name="5.11.0"></a>
# [5.11.0](https://github.com/softwaregroup-bg/ut-codec/compare/v5.10.1...v5.11.0) (2017-07-26)


### Features

* support multiple printers ([cbd1a98](https://github.com/softwaregroup-bg/ut-codec/commit/cbd1a98))



<a name="5.10.1"></a>
## [5.10.1](https://github.com/softwaregroup-bg/ut-codec/compare/v5.10.0...v5.10.1) (2017-07-18)


### Bug Fixes

* arqc format ([ffd1ce1](https://github.com/softwaregroup-bg/ut-codec/commit/ffd1ce1))



<a name="5.10.0"></a>
# [5.10.0](https://github.com/softwaregroup-bg/ut-codec/compare/v5.9.15...v5.10.0) (2017-07-18)


### Features

* zmk_arqc_arpc ([#8](https://github.com/softwaregroup-bg/ut-codec/issues/8)) ([1783665](https://github.com/softwaregroup-bg/ut-codec/commit/1783665))



<a name="5.9.15"></a>
## [5.9.15](https://github.com/softwaregroup-bg/ut-codec/compare/v5.9.14...v5.9.15) (2017-06-27)


### Bug Fixes

* dont expect method in response ([#7](https://github.com/softwaregroup-bg/ut-codec/issues/7)) ([b4d841a](https://github.com/softwaregroup-bg/ut-codec/commit/b4d841a))



<a name="5.9.14"></a>
## [5.9.14](https://github.com/softwaregroup-bg/ut-codec/compare/v5.9.13...v5.9.14) (2017-06-21)


### Bug Fixes

* jsonrpc full rewrite ([#6](https://github.com/softwaregroup-bg/ut-codec/issues/6)) ([76f97c8](https://github.com/softwaregroup-bg/ut-codec/commit/76f97c8))



<a name="5.9.13"></a>
## [5.9.13](https://github.com/softwaregroup-bg/ut-codec/compare/v5.9.12...v5.9.13) (2017-06-05)



<a name="5.9.12"></a>
## [5.9.12](https://github.com/softwaregroup-bg/ut-codec/compare/v5.9.11...v5.9.12) (2017-05-26)


### Bug Fixes

* generateKeyCheckValue - kcvType added to request pattern ([#5](https://github.com/softwaregroup-bg/ut-codec/issues/5)) ([aee9d63](https://github.com/softwaregroup-bg/ut-codec/commit/aee9d63))



<a name="5.9.11"></a>
## [5.9.11](https://github.com/softwaregroup-bg/ut-codec/compare/v5.9.10...v5.9.11) (2017-05-22)



<a name="5.9.10"></a>
## [5.9.10](https://github.com/softwaregroup-bg/ut-codec/compare/v5.9.9...v5.9.10) (2017-05-22)



<a name="5.9.9"></a>
## [5.9.9](https://github.com/softwaregroup-bg/ut-codec/compare/v5.9.8...v5.9.9) (2017-05-22)



<a name="5.9.8"></a>
## [5.9.8](https://github.com/softwaregroup-bg/ut-codec/compare/v5.9.7...v5.9.8) (2017-05-22)



<a name="5.9.7"></a>

tests/unit.emv.js
const test = require('blue-tape');
const emv = require('../emv');

const validEmv = {
    emvDolMsg: '8C159F02069F03069F1A0295055F2A029A039C019F37049F02060000000000009F03060000000000005A090000000000000000005F3401019F360200B69F2608322566E3E44A3DEA9F2701809F090200009F33036040209F1A0206089F350114570F0000000000000000000D27056209405F2A0206089A031707189F410215579C01019F3704000015579F53015A',
    emvConstructedTagMsg: '710A0102030405040708090A',
    emvLongLengthMsg: '5F0F8103AABBCC'
};

const invalidEmv = {
    emvConstructedTagMsgInvalidLength: {
        input: '710A0102030405060708090A',
        output: '710A0102030405040708090A'
    }
};

const integrityFailEmv = [
    '5F0F888803AABBCC',
    'qwerty'
];

test('Emv encode decode', (t) => {
    Object.keys(validEmv).forEach(k => {
        let decoded = emv.tagsDecode(validEmv[k], {});
        let decodedDols = emv.dolDecode(decoded);

        let encoded = emv.tagsEncode(decoded);

        let secondaryDecoded = emv.tagsDecode(encoded, {});
        let secondaryDecodedDols = emv.dolDecode(secondaryDecoded);

        t.deepEqual(decoded, secondaryDecoded, `ensure no data loss when encoding/decoding ${k}`);
        t.deepEqual(decodedDols, secondaryDecodedDols, `ensure no data loss when encoding/decoding ${k} with dols`);
    });

    Object.keys(invalidEmv).forEach(k => {
        let decoded = emv.tagsDecode(invalidEmv[k].input, {});
        let decodedDols = emv.dolDecode(decoded);

        let encoded = emv.tagsEncode(decoded);
        let encodedDols = emv.tagsEncode(decodedDols);

        t.deepEqual(encoded, invalidEmv[k].output, `ensure valid length calculation when encoding/decoding ${k}`);
        t.deepEqual(encodedDols, invalidEmv[k].output, `ensure valid length calculation when encoding/decoding ${k} with dols`);
    });

    integrityFailEmv.forEach(msg => {
        t.throws(() => { emv.tagsDecode(msg, {}); }, new Error('Data integrity error'), 'ensure invalid length throws error');
    });

    t.end();
});


tests/unit.ndc.js
var NDC = require('../ndc.js');
const ndc = new NDC({messageFormat: {}});
var test = require('blue-tape');
var emv = {
    request: {
        msg: new Buffer('31311c3030323030303030311c1c31383130303932321c31391c3b353032323635343030303030303030303d323730353632303934303f1c1c20444220202041441c3030303030303030303030301c31393634323d343f34393433393b313a1c1c1c1c32303032313130303030303030303030303030303030303030301c551c3543414d3034303038433135394630323036394630333036394631413032393530353546324130323941303339433031394633373034394630323036303030303030303030303030394630333036303030303030303030303030354130393530323236353430303030303030303030373546333430313031394633363032303042363946323630383332323536364533453434413344454139463237303138303946303930323030303039463333303336303430323039463141303230363038394633353031313435373046353032323635343030303030303030303037304432373035363230393430354632413032303630383941303331373037313839463431303231353537394330313031394633373034303030303135353739463533303135411c353833463239384503', 'hex'),
        result: { amount: '000000000000', bufferB: '', bufferC: '', camFlags: [ [ 0, 0, 1, 0, 0, 0, 0, 0 ], [ 0, 0, 0, 0, 0, 0, 0, 0 ] ], coordination: '9', emvTags: { '9F53': { len: 1, tag: '9F53', val: '5A' }, CDOL1: { len: 21, tag: '8C', val: { 95: { idx: 3, len: 5, tag: '95', val: '' }, amountAuthorised: { idx: 0, len: 6, tag: '9F02', val: '' }, amountOther: { idx: 1, len: 6, tag: '9F03', val: '' }, terminalCountryCode: { idx: 2, len: 2, tag: '9F1A', val: '' }, transactionCurrencyCode: { idx: 4, len: 2, tag: '5F2A', val: '' }, transactionDate: { idx: 5, len: 3, tag: '9A', val: '' }, transactionType: { idx: 6, len: 1, tag: '9C', val: '' }, unpredictableNumber: { idx: 7, len: 4, tag: '9F37', val: '' } } }, amountAuthorised: { len: 6, tag: '9F02', val: '000000000000' }, amountOther: { len: 6, tag: '9F03', val: '000000000000' }, applicationCryptogram: { len: 8, tag: '9F26', val: '322566E3E44A3DEA' }, applicationVersionNumber: { len: 2, tag: '9F09', val: '0000' }, atc: { len: 2, tag: '9F36', val: '00B6' }, cryptogramInformationData: { len: 1, tag: '9F27', val: '80' }, pan: { len: 9, tag: '5A', val: '502265400000000007' }, panSeqNum: { len: 1, tag: '5F34', val: '01' }, terminalCapabilities: { len: 3, tag: '9F33', val: '604020' }, terminalCountryCode: { len: 2, tag: '9F1A', val: '0608' }, terminalType: { len: 1, tag: '9F35', val: '14' }, track2EquivalentData: { len: 15, tag: '57', val: '5022654000000000070D2705620940' }, transactionCurrencyCode: { len: 2, tag: '5F2A', val: '0608' }, transactionDate: { len: 3, tag: '9A', val: '170718' }, transactionSequenceCounter: { len: 2, tag: '9F41', val: '1557' }, transactionType: { len: 1, tag: '9C', val: '01' }, unpredictableNumber: { len: 4, tag: '9F37', val: '00001557' } }, lastTransactionData: { notes1: 0, notes2: 0, notes3: 0, notes4: 0, sernum: '0021', status: '1' }, luno: '002000001', opcode: [ ' ', 'D', 'B', ' ', ' ', ' ', 'A', 'D' ], pinBlock: '19642D4F49439B1A', pinBlockNew: '', reserved: '', session: undefined, timeVariantNumber: '18100922', tokens: [ '11', '002000001', '', '18100922', '19', ';5022654000000000=2705620940?', '', ' DB   AD', '000000000000', '19642=4?49439;1:', '', '', '', '20021100000000000000000000', 'U', '5CAM04008C159F02069F03069F1A0295055F2A029A039C019F37049F02060000000000009F03060000000000005A095022654000000000075F3401019F360200B69F2608322566E3E44A3DEA9F2701809F090200009F33036040209F1A0206089F350114570F5022654000000000070D27056209405F2A0206089A031707189F410215579C01019F3704000015579F53015A', '583F298E\x03' ], topOfReceipt: '1', track2: ';5022654000000000=2705620940?', track3: '', type: '11' }
    },
    response: {
        msg: {
            'luno': '002000001',
            'coordination': '9',
            'timeVariantNumber': '19110959',
            'nextState': '165',
            'type1Notes': '00',
            'type2Notes': '00',
            'type3Notes': '00',
            'type4Notes': '00',
            'sernum': '0000',
            'function': '5',
            'screenUpdate': '',
            'screenMessage': '',
            'cardReturn': '0',
            'printer': '0',
            'printerData': '',
            'conId': 1,
            'session': {
                'transferId': false
            },
            'camFlags': [[0, 0, 0, 0, 0, 1, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0]],
            'emvTags': {
                // minimum required to encode
                'CDOL1': {
                    tag: '8C',
                    val: {
                        '95': {len: 5},
                        'amountAuthorised': {len: 6},
                        'amountOther': {len: 6},
                        'terminalCountryCode': {len: 2},
                        'transactionCurrencyCode': {len: 2},
                        'transactionDate': {len: 3},
                        'transactionType': {len: 1},
                        'unpredictableNumber': {len: 3}
                    }
                },
                'CDOL2': {
                    tag: '8D',
                    val: {
                        'transactionType': {len: 1},
                        'unpredictableNumber': {len: 3}
                    }
                },
                'amountAuthorised': {val: '000000000000'},
                'amountOther': {val: '000000000000'},
                'pan': {val: '502265400000000007'},
                'panSeqNum': {val: '01'},
                'atc': {val: '00B6'},
                'applicationCryptogram': {val: '322566E3E44A3DEA'},
                'cryptogramInformationData': {val: '80'},
                'applicationVersionNumber': {val: '0000'},
                'terminalCapabilities': {val: '604020'},
                'terminalCountryCode': {val: '0608'},
                'terminalType': {val: '14'},
                'track2EquivalentData': {val: '5022654000000000070D2705620940'},
                'transactionCurrencyCode': {val: '0608'},
                'transactionDate': {val: '170718'},
                'transactionSequenceCounter': {val: '1557'},
                'transactionType': {val: '01'},
                'unpredictableNumber': {val: '0000155F'},
                '9F53': {val: '5A'}
            }
        },
        result: '341c3030323030303030311c31393131303935391c3136351c30303030303030301c30303030351c3930303543414d30343030384331373935353946303236394630333639463141323546324132394133394331394633373338443439433139463337333946303236303030303030303030303030394630333630303030303030303030303035413935303232363534303030303030303030303735463334313031394633363230304236394632363833323235363645334534344133444541394632373138303946303932303030303946333333363034303230394631413230363038394633353131343537313535303232363534303030303030303030303730443237303536323039343035463241323036303839413331373037313839463431323135353739433130313946333734303030303135354639463533313541'
    }
};

test('Emv encode decode', (t) => {
    let resultReq = ndc.decode(emv.request.msg, {}, {});
    let resultResp = ndc.encode(emv.response.msg, {'auth': null, 'method': 'ncr.transactionReply', 'opcode': 'transactionReply', 'mtid': 'request', 'requestHeaders': {'host': 'localhost: 8006', 'accept': 'application/json', 'content-type': 'application/json', 'content-length': '384', 'connection': 'close'}, 'ipAddress': '127.0.0.1', 'conId': 1, 'trace': 'trn: 1'}, {'traceTransaction': 2, 'session': {'transferId': false}});

    t.deepEqual(resultReq, emv.request.result, 'decode');
    t.deepEqual(resultResp.toString('hex'), emv.response.result, 'encode');

    t.end();
});


emv.js
var emvTagsConfig = require('./emv.tags');
var emvTagsMap = emvTagsConfig.map;
var emvLongTags = emvTagsConfig.longTags;
var emvDolNumericTypes = emvTagsConfig.dolNumericTypes;

function emvEncodeMapTags() {
    emvTagsConfig.map.encode = Object.keys(emvTagsConfig.map.decode)
        .map((e) => {
            let o = {};
            o[emvTagsConfig.map.decode[e]] = e;
            return o;
        })
        .reduce((accum, cur) => Object.assign(accum, cur), {});
}

function translateTagEncode(tag) {
    return emvTagsMap.encode[tag] || tag;
}

function translateTagDecode(tag) {
    return emvTagsMap.decode[tag.toUpperCase()] || tag;
}

function getNumVal(val, len, dolLenDiff, compressedNumeric) {
    if (dolLenDiff < 0) { // left truncated
        return val.slice(len * -2);
    } else {
        if (!compressedNumeric) {
            return ((new Array(len)).fill('00').join('') + val).slice(len * -2);
        }
        return (val + (new Array(dolLenDiff)).fill('FF').join(''));
    }
};
function getNonNumVal(val, len, dolLenDiff) {
    if (dolLenDiff < 0) { // right truncated
        return val.slice(len * 2);
    } else {
        return (val + (new Array(len)).fill('00').join('')).slice(len * 2);
    }
};

/**
 * @param {Object} tagObj decoded emv tag
 * @param {string} tagObj.val decoded emv tag value
 * @param {string} tagObj.len decoded emv tag value bytes length (/2)
 */
function getValueHexLength(tagObj) {
    let len;
    if (!tagObj || (!tagObj.len && !tagObj.val)) {
        len = 0;
    } else {
        if (!tagObj.val || !tagObj.val.length) {
            len = tagObj.len || 0;
        } else {
            len = (tagObj.val.length / 2);
        }
    }
    len = len.toString(16).toUpperCase();

    return (len.length % 2) ? `0${len}` : len;
}

/**
 * @param {string} emvString emv input string
 * @param {Object} result object to store results in
 * @param {number=} dolIdx value position in data object list
 */
function tagsDecode(emvString, result, dolIdx) {
    let tag;
    let len;
    let val;
    let isDol = false;
    if (emvLongTags.includes(emvString.substr(0, 2).toLowerCase())) { // 2 bytes tag
        tag = emvString.substr(0, 4);
        emvString = emvString.substr(4);
    } else {
        tag = emvString.substr(0, 2);
        emvString = emvString.substr(2);
    }
    let tagTranslated = translateTagDecode(tag);
    result[tagTranslated] = {tag};
    if (tagTranslated.includes('DOL')) {
        isDol = true;
    }
    if (dolIdx) {
        result[tagTranslated].idx = dolIdx - 1;
    }
    let lenStr = emvString.substr(0, 2);
    len = (lenStr === '') ? 0 : parseInt(emvString.substr(0, 2), 16);
    emvString = emvString.substr(2);
    if (len >= 128) { // size is big
        let byteNumSize = 0;
        let cur = 128;
        while (cur >= 1) { // calculate big size
            cur = cur >> 1;
            if ((len & cur) === cur) {
                byteNumSize = byteNumSize | cur;
            }
        }
        len = parseInt(emvString.substr(0, byteNumSize * 2), 16);
        emvString = emvString.substr(byteNumSize * 2);
    }
    if (!dolIdx && len > emvString.length / 2) {
        throw new Error('Data integrity error');
    }
    result[tagTranslated].len = len;
    if (!len) {
        len = 0;
        val = '';
    } else {
        if (dolIdx) {
            val = '';
        } else {
            val = emvString.substr(0, len * 2);
            emvString = emvString.substr(len * 2);
        }
    }
    let constructedTagByte = (new Buffer(tag, 'hex')).slice(0, 1)[0];
    if ((constructedTagByte & 32) === 32) {
        result[tagTranslated].val = (isDol ? tagsDecode(val, {}, 1) : tagsDecode(val, {}, (dolIdx ? dolIdx + 1 : dolIdx)));
    } else {
        result[tagTranslated].val = (isDol ? tagsDecode(val, {}, 1) : val);
    }
    if (emvString.length) {
        return tagsDecode(emvString, result, (dolIdx ? dolIdx + 1 : dolIdx));
    }
    return result;
};

/**
 * EMV 4.3 Book 3                              5 Data Elements and Files
 * Application Specification                   5.4 Rules for Using a Data Object List (DOL)
 * @param {Object} emvTags decoded emv tags map
 */
function dolDecode(emvTags) {
    let data = Object.assign({}, emvTags);
    let mainTags = Object.keys(data);
    let dolTags = mainTags.filter((t) => (t.includes('DOL')));
    if (dolTags.length) {
        data = dolTags
            .map((t) => ({tag: t, data: data[t].val, internalTags: Object.keys(data[t].val)}))
            .reduce((allTags, dol) => {
                allTags[dol.tag].val = dol.internalTags.reduce((dolTags, dolInt) => {
                    let extTag = allTags[dolInt];
                    let dolTag = dolTags[dolInt];
                    if (!extTag) { // no tag found in root tags list
                        dolTags[dolInt].val = (new Array(dolTags[dolInt].len)).fill('00').join('');
                    } else if (extTag.len === dolTag.len) {
                        dolTags[dolInt].val = extTag.val;
                    } else {
                        let extNumType = emvDolNumericTypes[dolTag.tag];
                        if (extNumType === 'n') { // non numeric value
                            dolTags[dolInt].val = getNonNumVal(extTag.val, dolTag.len, dolTag.len - extTag.len, false);
                        } else if (extNumType === 'nc') { // non numeric value
                            dolTags[dolInt].val = getNonNumVal(extTag.val, dolTag.len, dolTag.len - extTag.len, true);
                        } else { // numeric value
                            dolTags[dolInt].val = getNumVal(extTag.val, dolTag.len, dolTag.len - extTag.len);
                        }
                    }
                    return dolTags;
                }, dol.data);
                return allTags;
            }, data);
    }
    return data;
};

/**
 * @param {Object} emvTags decoded emv tags map
 */
function tagsEncode(emvTags) {
    let data = Object.assign({}, emvTags);
    let dolOrder = ['CDOL1', 'CDOL2', 'TDOL', 'PDOL', 'DDOL'];
    let result = '';
    let allTags = Object.keys(data)
        .sort((a, b) => {
            if((!data[a].idx && data[a].idx !==0) || (!data[b].idx && data[b].idx !== 0)) {
                return 0;
            }
            return data[a].idx - data[b].idx;
        });
    // transform data in dols
    data = allTags
        .filter((k) => (k.includes('DOL')))
        .map((k) => {
            let dol = data[k];
            return {
                tag: dol.tag,
                data: Object.keys(dol.val)
                    .sort((a, b) => {
                        return dol.val[a].idx > dol.val[b].idx;
                    })
                    .map((k) => {
                        let tagObj = dol.val[k];
                        let tagTranslated = translateTagEncode(k);
                        let tagLength = getValueHexLength(tagObj);

                        return [tagTranslated, tagLength].join('');
                    })
            };
        })
        .reduce((data, dol) => {
            data[dol.tag] = dol.data.join('');
            return data;
        }, data);
    // make sure that dols are constructed in order
    let allDols = dolOrder
        .map((dol) => (allTags.includes(dol) ? dol : 0))
        .filter((e) => e);
    // append dols to result
    result = allDols
        .reduce((r, dol) => {
            let tagTranslated = translateTagEncode(dol);
            let d = Object.assign({}, data[dol], {val: data[tagTranslated]});
            let tagLength = getValueHexLength(d);

            return `${r}${tagTranslated}${tagLength}${d.val}`;
        }, '');
    // cleanup dols
    data = allDols
        .reduce((r, dol) => {
            let tagTranslated = translateTagEncode(dol);
            delete r[dol];
            delete r[tagTranslated];
            return r;
        }, data);
    // append all fields left to result
    return result + allTags.map((e) => {
        let tagTranslated = translateTagEncode(e);
        let tagObj = data[e];
        let tagLength = getValueHexLength(tagObj);
        let tagValue = tagObj.val;

        let constructedTagByte = (new Buffer(tagTranslated, 'hex')).slice(0, 1)[0];
        if ((constructedTagByte & 32) === 32) {
            tagValue = tagsEncode(tagObj.val);
        }

        return `${tagTranslated}${tagLength}${tagValue}`;
    }).join('');
}

module.exports = (function emv() {
    if (!emvTagsConfig.map.encode) {
        emvEncodeMapTags();
    }

    return {
        tagsDecode,
        dolDecode,
        tagsEncode
    };
})();


iso8583.js
'use strict';
var merge = require('lodash.merge');
var defaultFields = require('./iso8583.fields.json');
var bitSyntax = require('ut-bitsyntax');
var errors = require('./isoErrors');
var emv = require('./emv');

function getFormat(format, fallback) {
    return (format && {'numeric': 'string-left-zero', 'string': 'string-right-space', 'amount': 'string-left-zero', 'bcdamount': 'string'}[format]) || format || fallback || 'binary';
}

function Iso8583(config) {
    this.emvTagsField = config.emvTagsField || 55;
    this.fieldFormat = merge({}, defaultFields[(config.version || '0') + (config.baseEncoding || 'ascii')], config.fieldFormat);
    this.framePattern = bitSyntax.matcher('header:' + this.fieldFormat.header.size + '/' + getFormat(this.fieldFormat.header.format) +
        ', mtid:' + this.fieldFormat.mtid.size + '/' + getFormat(this.fieldFormat.mtid.format) +
        ', field0:' + this.fieldFormat['0'].size + '/' + getFormat(this.fieldFormat['0'].format) +
        ', rest/binary');
    this.fieldPatterns = [];
    this.fieldBuilders = [bitSyntax.parse('field:fieldSize/' + getFormat(this.fieldFormat['0'].format))];
    this.fieldBuilders.mtid = bitSyntax.parse('field:fieldSize/' + getFormat(this.fieldFormat.mtid.format));
    this.fieldBuilders.header = bitSyntax.parse('field:fieldSize/' + getFormat(this.fieldFormat.header.format));
    this.fieldBuilders.footer = bitSyntax.parse('field:fieldSize/' + getFormat(this.fieldFormat.footer.format));
    this.prefixBuilders = [null];
    this.footerMatcher = bitSyntax.matcher('footer:' + this.fieldFormat.footer.size + '/' + getFormat(this.fieldFormat.footer.format) + ', rest/binary');
    var group = 0;
    while (this.fieldFormat[(group + 1) * 64]) {
        var pattern = [];
        for (var i = 1; i <= 64; i += 1) {
            var field = group * 64 + i;
            if (this.fieldFormat[field].prefixSize) { // if the field is with variable size
                pattern.push('prefix' + field + ':field' + field + 'Size/' + getFormat(this.fieldFormat[field].prefixFormat, 'string-left-zero') +
                    ', field' + field + ':prefix' + field + '/' + getFormat(this.fieldFormat[field].format));
                this.prefixBuilders.push(bitSyntax.parse('prefix:' + this.fieldFormat[field].prefixSize + '/' +
                    getFormat(this.fieldFormat[field].prefixFormat, 'string-left-zero')));
            } else { // if the field is with fixed size
                pattern.push('field' + field + ':field' + field + 'Size/' + getFormat(this.fieldFormat[field].format));
                this.prefixBuilders.push(null);
            }
            this.fieldBuilders.push(bitSyntax.parse('field:fieldSize/' + getFormat(this.fieldFormat[field].format)));
        }
        pattern.push('rest/binary');
        this.fieldPatterns.push(bitSyntax.matcher(pattern.join(', ')));
        group += 1;
    }
}

Iso8583.prototype.fieldSizes = function(bitmap, start) {
    /* jshint bitwise: false */
    var result = {};
    for (var i = 0; i <= 63; i += 1) {
        var size = bitmap && ((bitmap[i >> 3] & (128 >> (i % 8))) !== 0);
        if (size) {
            result['field' + (start + i) + 'Size'] = this.fieldFormat[start + i].prefixSize || this.fieldFormat[start + i].size;
        } else {
            result['field' + (start + i) + 'Size'] = 0;
        }
    }
    return result;
};

Iso8583.prototype.decode = function(buffer, $meta) {
    var frame = this.framePattern(buffer);
    var bitmapField = 0;
    if (frame) {
        var message = {'header': frame.header, 'mtid': frame.mtid, '0': frame.field0};
        var parsedLength = buffer.length - frame.rest.length;
        var group = 0;
        while (frame) {
            var fieldPattern = this.fieldPatterns[group];
            if (!fieldPattern) {
                frame = this.footerMatcher(frame.rest);
                message.footer = frame && frame.footer;
                if (frame.rest && frame.rest.length) {
                    throw new Error('Not all data was parsed. Remaining ' + frame.rest.length + ' bytes at offset ' + parsedLength +
                        ' starting with 0x' + frame.rest.toString('hex') + '\r\nmessage:' + JSON.stringify(message));
                } else {
                    break;
                }
            }
            var fieldSizes = this.fieldSizes(frame['field' + bitmapField], group * 64 + 1);
            var rest = frame.rest;
            frame = fieldPattern && fieldPattern(rest, fieldSizes);
            if (!frame && fieldPattern) {
                for (var failField = (group + 1) * 64; failField >= group * 64 + 1; failField -= 1) { // find at which field we failed by skipping fields from the end
                    fieldSizes['field' + failField + 'Size'] = 0;
                    frame = fieldPattern && fieldPattern(rest, fieldSizes);
                    if (frame) {
                        frame = this.footerMatcher(frame.rest);
                        message.footer = frame && frame.footer;
                    }
                    if (frame) {
                        parsedLength += rest.length - frame.rest.length;
                        throw new Error('Parsing failed at field ' + failField + '. Remaining ' + frame.rest.length + ' bytes at offset ' + parsedLength +
                            ' starting with 0x' + frame.rest.toString('hex') + '\r\nmessage:' + JSON.stringify(message));
                    }
                }
                throw new Error('Parsing failed at unknown field');
            }
            parsedLength += rest.length - frame.rest.length;
            bitmapField = group * 64 + 1;
            for (var fieldNo = group * 64 + 1; fieldNo <= (group + 1) * 64; fieldNo += 1) {
                if (fieldSizes['field' + fieldNo + 'Size']) {
                    message[fieldNo] = frame['field' + fieldNo];
                }
            }
            group += 1;
        }
        if (message.mtid === '0800') {
            $meta.opcode = String(message[70] || '').substr(0, 3);
        } else {
            $meta.opcode = String(message[3] || '').substr(0, 2);
        }
        $meta.trace = message[11];
        if (message.mtid && message.mtid.slice) {
            $meta.mtid = {
                '0': 'request',
                '1': (parseInt(message[39]) === 0) ? 'response' : 'error',
                '2': 'request',
                '3': (parseInt(message[39]) === 0) ? 'response' : 'error',
                '4': 'notification',
                '5': 'notification'
            }[(message.mtid.slice(-2).substr(0, 1))] || 'error';
        }
        $meta.method = message.mtid + ($meta.opcode ? '.' + $meta.opcode : '');
        if ($meta.mtid === 'error') {
            var err = errors['' + message[39]] || errors.generic;
            message = err(message);
        }
        if (message[this.emvTagsField]) {
            message = Object.assign(message, {emvTags: emv.tagsDecode(message[this.emvTagsField])});
        }
        return message;
    } else {
        throw new Error('Unable to parse message type or first bitmap!');
    }
};

Iso8583.prototype.encodeField = function(fieldName, fieldValue) {
    var prefixBuilder = this.prefixBuilders[fieldName];
    var field = bitSyntax.build(
        this.fieldBuilders[fieldName],
        {'field': fieldValue, 'fieldSize': prefixBuilder ? fieldValue.length : this.fieldFormat[fieldName].size}
    );
    return prefixBuilder ? Buffer.concat([bitSyntax.build(prefixBuilder, {'prefix': field.length}), field]) : field;
};

Iso8583.prototype.encode = function(message, $meta, context) {
    /* jshint bitwise: false */
    var buffers = new Array(64 * this.fieldPatterns.length);
    var emptyBuffer = new Buffer([]);
    var trace = message[11];
    if (message.emvTags) {
        message[this.emvTagsField] = emv.encode(message.emvTags);
    }
    if (trace === undefined || trace === null) {
        trace = $meta.trace = ('000000' + context.trace).substr(-6);
        context.trace += 1;
        if (context.trace > 999999) {
            context.trace = 0;
        }
    }
    message[11] = trace;
    var bitmaps = Array.apply(null, new Array(8 * this.fieldPatterns.length)).map(Number.prototype.valueOf, 0); // zero filled array
    for (var i = 64 * this.fieldPatterns.length; i >= 0; i -= 1) {
        if (i === 0) {
            buffers[i] = this.encodeField(i, new Buffer(bitmaps.slice(0, 8)));
        } else if (i % 64 === 1 && i < 64 * (this.fieldPatterns.length - 1)) {
            var index = (i >> 6) << 3;
            var bitmap = bitmaps.slice(index + 8, index + 16);
            if (bitmap.reduce(function(p, n) { return p + n; })) {
                bitmaps[(i - 1) >> 3] |= (128 >> (i - 1) % 8);
                buffers[i] = this.encodeField(i, new Buffer(bitmap));
            } else {
                buffers[i] = emptyBuffer;
            }
        } else if (message[i] != null) {
            bitmaps[(i - 1) >> 3] |= (128 >> (i - 1) % 8);
            buffers[i] = this.encodeField(i, message[i]);
        } else {
            buffers[i] = emptyBuffer;
        }
    }
    buffers.unshift(this.encodeField('mtid', message.mtid || new Buffer([])));
    buffers.unshift(this.encodeField('header', message.header || Buffer.alloc(this.fieldFormat.header.size)));
    buffers.push(this.encodeField('footer', message.footer || Buffer.alloc(this.fieldFormat.footer.size)));

    return Buffer.concat(buffers);
};

module.exports = Iso8583;
